src/mds/MDLog.cc

```diff
@@ -1314,6 +1328,12 @@ void MDLog::_replay_thread()
 	   !journaler->get_error()) {
       C_SaferCond readable_waiter;
       journaler->wait_for_readable(&readable_waiter);
+	   
+      mutex.lock();
+      timer_checker = new C_Replay_Checker(this);
+      init_timer.add_event_after(10.0, timer_checker );
```

Comment:
I don't think this is the right approach to fixing this issue. Have you thought about serializing via mutexes?

Reply:
Code1:
    if (on_readable) {
      C_OnFinisher *f = on_readable;
      on_readable = 0;
      f->complete(error);
    }
	
Code2:
    while (!journaler->is_readable() &&
	   journaler->get_read_pos() < journaler->get_write_pos() &&
	   !journaler->get_error()) {
      C_SaferCond readable_waiter;
      journaler->wait_for_readable(&readable_waiter);
      r = readable_waiter.wait();
    }
	
Code1 is called in Journaler::_finish_read(), Journaler::_assimilate_prefetch(), Journaler::shutdown().
Code2 is called in MDLog::_replay_thread(), MDLog::_reformat_journal().

If we use mutexes to serialize it, maybe we should:
    (1)add lock before Code1 and Code2;
    (2)add unlock after Code1 and after journaler->wait_for_readable() in Code2;
Is that the right idea to process it?

It is a better and more complex method:
     _finish_read() is called by is_readable()->_prefetch()->_issue_read(), but _prefetch() and is_readable() is called in multiple different function call stack. We will need to spend more time checking its correctness.

Reply:
> I don't think this is the right approach to fixing this issue. Have you thought about serializing via mutexes?
@vshankar  After solving above deadlock problem, mutex method can work, I had tested it for this issue. This is a mutex version patch (see attachment).

((Files with ".patch" as the suffix cannot be uploaded. I added ".log".))
[replaythread_hanged_mutex .patch.log](https://github.com/ceph/ceph/files/9738313/replaythread_hanged_mutex.patch.log)

Reply:
> > I don't think this is the right approach to fixing this issue. Have you thought about serializing via mutexes?
> > After solving above deadlock problem, mutex method can work, I had tested it for this issue.
> > I had change the commit.

I am sorry to tell you that: this mutex version has a problem, I will give a new version later. @vshankar

Reply:
Sure. Let me know when its ready for review.

Reply:
There is a new version already, we will review it Internally and make more tests.
I think I can provide it tomorow.

Reply:
@vshankar The new version is ready for review.
  (1) I add a new C_SaferCond_ReferenceLock. In the C_SaferCond, the lock is locked in the C_SaferCond::wait, I cannot use it to ensure C_SaferCond::complete() is called afer C_SaferCond::wait(). 
  (2) I add a Journaler::replaylock. If I use the Journaler::lock, I should delete the lock used in is_readable() and wait_for_readable(), but is_readable() has been called several times in other function.

Reply:
@vshankar  If we donot want to introduce new class C_SaferCond_ReferenceLock, we can modify Journaler::check_isreadable():
<img width="536" alt="mutex1" src="https://user-images.githubusercontent.com/44566632/195240166-0d552938-4ef0-4ce5-87d7-e11ed651b8c5.png">
I have tested it and it can work well too.

In this modification, if replay_thread switch to finisher_thread just before "r=readable_waiter.wait()", and "finish_read and f->complete()" is executed before "r=readable_waiter.wait()", the "r=readable_waiter.wait()" will not block replay_thread.

If there is partial fragment in "_finish_read->_assimilate_prefetch->_is_readable()", the write_pos will equal to read_pos; or the is_readable() will return true. It will not enter the inner while again.